File: code/common/tabular.py
"""Reading and writing the pipeline's tables. Mirrors code/common/tabular.R.
"""

from __future__ import annotations

import re
from pathlib import Path

import pandas as pd

_ID_COLUMNS = ("eid", "participant_id", "pmr_id", "ref_id")
_NA_VALUES = ["NA"]
_ISO_DATE = re.compile(r"^\d{4}-\d{2}-\d{2}$")


def _restore_dates(frame: pd.DataFrame) -> pd.DataFrame:
    for col in frame.columns:
        if frame[col].dtype != object:
            continue
        seen = frame[col].dropna()
        if seen.empty or not all(isinstance(v, str) for v in seen.head(1000)):
            continue
        if seen.map(lambda v: bool(_ISO_DATE.match(str(v)))).all():
            frame[col] = pd.to_datetime(frame[col], format="%Y-%m-%d")
    return frame


def read_table(path, usecols=None, nrows=None, dtype=None, **kw):
    kw.pop("na", None)
    kw.pop("low_memory", None)
    p = Path(path)
    if not p.exists():
        raise FileNotFoundError(f"No table at {p}")
    if p.suffix == ".xlsx":
        frame = pd.read_excel(p, usecols=list(usecols) if usecols else None,
                              nrows=int(nrows) if nrows is not None else None,
                              dtype=dtype, na_values=_NA_VALUES,
                              keep_default_na=False)
        return frame
    frame = pd.read_csv(
        p,
        usecols=list(usecols) if usecols else None,
        nrows=int(nrows) if nrows is not None else None,
        dtype=dtype,
        na_values=_NA_VALUES,
        keep_default_na=False,
        low_memory=False,
        **kw,
    )
    if dtype is None:
        frame = _restore_dates(frame)
    widen = {c: "int64" for c in _ID_COLUMNS
             if c in frame.columns and str(frame[c].dtype).startswith("int")}
    if widen:
        frame = frame.astype(widen)
    return frame


def write_table(frame: pd.DataFrame, path, index: bool = False) -> Path:
    p = Path(path)
    p.parent.mkdir(parents=True, exist_ok=True)
    out = frame if index else frame.reset_index(drop=True)
    if p.suffix == ".xlsx":
        out.to_excel(p, index=index, na_rep="NA")
    else:
        out.to_csv(p, index=index, na_rep="NA", date_format="%Y-%m-%d")
    return p

File: code/common/tabular.R
# Reading and writing the pipeline's tables. Mirrors code/common/tabular.py.

.ID_COLUMNS <- c("eid", "participant_id", "pmr_id", "ref_id")
.ISO_DATE <- "^[0-9]{4}-[0-9]{2}-[0-9]{2}$"

.as_date <- function(days) structure(as.numeric(days), class = "Date")

.restore_dates <- function(out) {
  for (col in names(out)) {
    x <- out[[col]]
    if (inherits(x, "IDate")) {
      out[[col]] <- .as_date(unclass(x))
      next
    }
    if (!is.character(x)) next
    seen <- x[!is.na(x)]
    if (!length(seen)) next
    if (all(grepl(.ISO_DATE, seen))) out[[col]] <- .as_date(unclass(as.Date(x)))
  }
  out
}

read_table <- function(path, select = NULL, ...) {
  if (!file.exists(path)) stop("No table at ", path, call. = FALSE)
  if (grepl("\\.xlsx$", path)) {
    out <- as.data.frame(readxl::read_excel(path), stringsAsFactors = FALSE)
    if (!is.null(select)) out <- out[, select, drop = FALSE]
    return(out)
  }
  args <- list(...)
  args$file <- path
  args$data.table <- FALSE
  args$showProgress <- FALSE
  args$stringsAsFactors <- NULL
  if (is.null(args$na.strings)) args$na.strings <- "NA"
  args$keepLeadingZeros <- TRUE
  if (!is.null(select)) args$select <- select
  out <- as.data.frame(do.call(data.table::fread, args), stringsAsFactors = FALSE)
  out <- .restore_dates(out)
  for (col in intersect(.ID_COLUMNS, names(out))) {
    if (is.integer(out[[col]])) out[[col]] <- as.numeric(out[[col]])
  }
  out
}

write_table <- function(x, path) {
  dir.create(dirname(path), recursive = TRUE, showWarnings = FALSE)
  x <- as.data.frame(x)

  if (grepl("\\.xlsx$", path)) {
    writexl::write_xlsx(x, path)
    return(invisible(path))
  }

  for (col in names(x)) {
    v <- x[[col]]
    if (is.factor(v)) v <- levels(v)
    if (is.character(v) && any(grepl('"', v, fixed = TRUE))) {
      stop("Column ", col, " of ", basename(path), " holds a double quote, ",
           "which this format cannot carry.", call. = FALSE)
    }
    if (is.double(x[[col]]) && !inherits(x[[col]], "Date")) {
      x[[col]] <- ifelse(is.na(x[[col]]), NA_character_,
                         sprintf("%.17g", x[[col]]))
    }
  }
  data.table::fwrite(x, path, na = "NA", quote = "auto", dateTimeAs = "ISO",
                     logical01 = FALSE)
  invisible(path)
}
